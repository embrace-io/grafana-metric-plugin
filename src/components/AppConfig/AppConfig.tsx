import React, {FC, useCallback, useEffect, useState} from 'react';
import {AppPluginMeta, PluginConfigPageProps} from '@grafana/data';
import {getBackendSrv} from '@grafana/runtime';
import {Button, InlineField, Input} from '@grafana/ui';
import {ApplicationName, ApplicationRoot} from '../../constants';
import {GlobalSettings} from '../../types';

/**
 * Page Properties
 */
export interface Props extends PluginConfigPageProps<AppPluginMeta<GlobalSettings>> {
}

/**
 * Config component
 */
export const Config: FC<Props> = (props) => {
    const [isEnabled, setIsEnabled] = useState(!!props.plugin.meta?.enabled);
    const [apiToken, setApiToken] = useState('');
    const [validToken, setValidToken] = useState(false);
    const [showErrorMessage, setShowErrorMessage] = useState(false);

    /**
     * Plugin Settings
     *
     * @param settings Plugin Settings
     */
    const updatePluginSettings = useCallback((settings: { enabled: boolean; jsonData: unknown; pinned: boolean }): Promise<undefined> => {
        return getBackendSrv().post(`api/plugins/${props.plugin.meta.id}/settings`, settings);
    }, [props]);

    /**
     * Plugin disable
     */
    const onDisable = () => {
        updatePluginSettings({enabled: false, jsonData: {}, pinned: false}).then(() => {
            window.location.reload();
        });
    };

    /**
     * Plugin enable
     */
    const onEnable = async () => {
        let dataSourceInfo;

        // Check to see if Embrace Metrics API datasource exists already
        await getBackendSrv()
            .get('/api/datasources').then((results) => {
                results.forEach((dataSource: { name: string; uid: string; }) => {
                    if (dataSource.name === 'Embrace Metrics API') {
                        dataSourceInfo = dataSource;
                    }
                })
            });

        // If it does not exist not, create it
        if (dataSourceInfo === undefined && !showErrorMessage) {
            const results = await getBackendSrv()
                .post('/api/datasources', {
                    name: "Embrace Metrics API",
                    type: "prometheus",
                    access: 'proxy',
                });
            dataSourceInfo = results.datasource;
        }

        // Make PUT request configuring the URL and Authorization header for the datasource
        await getBackendSrv()
            .put(`api/datasources/uid/${dataSourceInfo.uid}`, {
                ...dataSourceInfo,
                url: 'https://api.embrace.io/metrics',
                jsonData: {httpMethod: 'POST', httpHeaderName1: 'Authorization'},
                secureJsonData: {httpHeaderValue1: 'Bearer ' + apiToken},
            });

        // Test if token is valid
        await getBackendSrv().get(`api/datasources/proxy/${dataSourceInfo.id}/api/v1/status/buildinfo`).catch((error) => {
            if (error.status === 405) {
                setValidToken(true);
            } else {
                setShowErrorMessage(true);
            }
        })
    };

    useEffect(() => {
        if (validToken) {
            updatePluginSettings({enabled: true, jsonData: {}, pinned: true}).then(() => {
                setIsEnabled(true);
                window.location.assign(ApplicationRoot);
            });
        }
    }, [validToken, updatePluginSettings]);

    return (
        <React.Fragment>
            <h2>{ApplicationName}</h2>
            <p>The Embrace App is a plugin for Grafana that gives teams an easy way to visualize Embrace dashboards using Embrace Metrics API data powered by Prometheus.</p>
            {!isEnabled && (
                <React.Fragment>
                    <p>
                        Get out-of-the-box Embrace dashboards in minutes with our simple configuration. Enter your Embrace API Token below and click to Enable the Application.
                    </p>
                    <InlineField
                        label="Embrace API Token"
                        error="Your API token was invalid. Please try again"
                        invalid={showErrorMessage}
                    >
                        <Input placeholder="Token"
                               onChange={(e) => setApiToken(e.currentTarget.value)}
                               value={apiToken}
                               width={50}
                               />
                    </InlineField>
                </React.Fragment>
            )}
            <div className="gf-form gf-form-button-row">
                {isEnabled ? (
                    <Button variant="destructive" onClick={onDisable}>
                        Disable
                    </Button>
                ) : (
                    <Button onClick={onEnable} disabled={apiToken === ''}>Enable</Button>
                )}
            </div>
        </React.Fragment>
    );
}
