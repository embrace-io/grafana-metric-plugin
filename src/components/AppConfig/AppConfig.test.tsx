import React from 'react';
import { render } from '@testing-library/react';
import { PluginType } from '@grafana/data';
import AppConfig, { type AppConfigProps } from './AppConfig';

describe('Components/AppConfig', () => {
  let props: AppConfigProps;

  beforeEach(() => {
    jest.resetAllMocks();

    props = {
      plugin: {
        meta: {
          id: 'sample-app',
          name: 'Sample App',
          type: PluginType.app,
          enabled: false,
          jsonData: {},
        },
      },
      query: {},
    } as unknown as AppConfigProps;
  });

  test('renders without an error"', () => {
    const screen = render(<AppConfig plugin={props.plugin} query={props.query} />);
    expect(screen.container).toMatchSnapshot();
  });

  test('renders an "Enable" button if the plugin is disabled', () => {
    const plugin = { meta: { ...props.plugin.meta, enabled: false } };

    // @ts-ignore - We don't need to provide `addConfigPage()` and `setChannelSupport()` for these tests
    const screen = render(<AppConfig plugin={plugin} query={props.query} />);

    expect(screen.queryByText('Enable')).toBeInTheDocument();
    expect(screen.queryByText('Disable')).not.toBeInTheDocument();
  });

  test('renders a "Disable" button if the plugin is enabled', () => {
    const plugin = { meta: { ...props.plugin.meta, enabled: true } };

    // @ts-ignore - We don't need to provide `addConfigPage()` and `setChannelSupport()` for these tests
    const screen = render(<AppConfig plugin={plugin} query={props.query} />);

    expect(screen.queryByText('Disable')).toBeInTheDocument();
    expect(screen.queryByText('Enable')).not.toBeInTheDocument();
  });
});
