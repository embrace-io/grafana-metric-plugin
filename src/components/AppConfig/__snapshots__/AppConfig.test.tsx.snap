// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/AppConfig renders without an error" 1`] = `
<div>
  <h2>
    Embrace
  </h2>
  <p>
    The Embrace App is a plugin for Grafana that gives teams an easy way to visualize Embrace dashboards using Embrace Metrics API data powered by Prometheus.
  </p>
  <p>
    Get out-of-the-box Embrace dashboards in minutes with our simple configuration. Enter your Embrace API Token below and click to Enable the Application.
  </p>
  <div
    class="css-1f5gueu"
  >
    <label
      class="css-1pews2r"
    >
      Embrace API Token
    </label>
    <div
      class="css-1rfunm5"
    >
      <div
        class="css-p7rfez-input-wrapper"
        data-testid="input-wrapper"
      >
        <div
          class="css-1w5c5dq-input-inputWrapper"
        >
          <input
            class="css-1wdli31-input-input"
            placeholder="Token"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="gf-form gf-form-button-row"
  >
    <button
      class="css-1sara2j-button"
      disabled=""
      type="button"
    >
      <span
        class="css-1mhnkuh"
      >
        Enable
      </span>
    </button>
  </div>
</div>
`;
